### DISCUSSION TITLE
Creating you own LUT

### DISCUSSION THREAD

**Discussion by Bascht74** [awawa-dev/HyperHDR#35] [Q&A]
Hello, 

when I use KODI on my FireCube in 4K/HDR to render a blue image (0,0,255) it comes out a bit too red (47,8,255) after applying the default HDR LUT. 
That is not good with SK6812, because the blue it not the greatest/cleanest...
I played around with the HUE setting, but that does no good with all other colors...

I have found the possibility to generate a LUT in HyperHDR. 
But there seems to be no description/documentation for this, at least I didn't find one. 

Do you have some info on how to use that tool inside HyperHDR?
Do you need a special image that you display with Kodi, for example, and then record again via the HDMI grabber?

Thanks (again :-) for any advice...

Sebastian

**Comment by Bascht74** [comment]
Funny thing is, that if I am calibrating the image and I am at the red picture and enter rgb(0,0,0) = black, the leds a not turning off, but they turn very dark yellow...
Same with green and enter rgb(0,0,0) it won't turn black.

And with the blue test picture and entering rgb(0,0,0) the leds won't turn off but turn dark red/pinkish...
That is strange... It seems that you cannot control the color of the leds completly...
The red pinkish is what I want to get rid off...
I tried negative values (because 0,0,0 is not 0,0,0), but it won't accept it. 

Could you quickly check that with your setup?

I have Pi4 with SK6812 RGBW and your ESP8266 firmware for the white calibration...

**Reply by awawa-dev** [reply]
Generic LUT table has to work with a wild range of different grabbers/splitter that do some damage to HDR->SDR processing. It's a kind of rebuilding HDR from SDR again using mix of algorithms. It's no perfect method but it works well in most cases. The weakness is mostly seen on dark scenes and you can workaround it to some extent with calibration in processing tab (but it's got a limitation because it's not a linear problem). There is uploaded dedicated LUT table for Ezcap 320 and it provides much better results for HDR for 1080p for that model but it's a manual work with many trials and experiments to create one on your own.

**Reply by Bascht74** [reply]
Hi @awawa-dev 

can you explain a bit more how to create a HDR -> SDR LUT for the tone mapping itself?

I bought the "usual" cheap HDMI grabber, which basically works fine:
https://www.amazon.de/gp/product/B08DFRS36T/

However, with HDR video, too much red is mixed to the blue, even though I set 0,0,255 for blue. So this already happens during tone mapping and can also be seen very well via the web interface of hyperHDR (if you hold a color picker on the pure blue of KODI it shows rgb 40,0,255). 

I would like to create my own LUT for the above caputure card, which I would be happy to share with others. 

However, I would need some information for this, e.g. what format the .3d file must have and some rough steps on how to proceed (e.g. how to get a test image directly from the card, can HyperHDR save a screenshot? which tool to use to create it).

Greetings and thanks for your help!

Sebastian

**Reply by Bascht74** [reply]
I found the file lut_lin_tables.3d in /usr/share/hyperhdr/lut/ 
Maybe you can change the documentation in the web interface...


> HDR to SDR tone mapping | If enabled, HyperHDR will try to correct colors of the HDR10 content that was captured by the grabber as SDR format. 3D LUT file 'lut_lin_tables.3d' is already included but if you want you own please put it in HyperHDR configuration folder.You can generate one and preview the effect on that page: LINK. That file is also strongly recommended for YUV grabbers even if you don't use HDR.
> **Typical configuration folder for is (check logs in DEBUG mode): Rpi /home/pi/.hyperhdrWindows c:/Users/NAME/.hyperhdr**

**Reply by awawa-dev** [reply]
The folders from doc are still valid and if you put a lut file inside them (for example my custom LUT for Ezcap 320) that file has a higher priority than the standard LUT file that you have found in /usr/share/hyperhdr/lut/ . Format of LUT file is internal but you can edit  javascipt of the generator page to put some correction on color channel and generate one that will fits your needs.

**Reply by Bascht74** [reply]
@awawa-dev Thank you for you answer! 
Where can I find the custom LUT for the Ezcap 320? I have looked everywhere so maybe I am too stupid to find it or I am overseeing it somehow.

Btw.: I found a solution for my "too red" blues: I changed the hue a little bit (+4) and they are much better....

The only thing I have not found out is that dark colors are too strong. They look very dark in the LED-Visualisation but on the wall they are very pure (too pure). Brighter Colors are OK. 

Could you probably insert an option to reduce the color purity for dark colors (mix more white in, keep the brightness)? 
If I understand right, this could be done best in the esp8266-code, right? Maybe here (I am not a programmer, so hopefully my logic is right): 
https://github.com/awawa-dev/HyperSerialEsp8266/blob/43e484832da82f77c9734d5b73bcf3cf5c4ef517/HyperSerialEsp8266.ino#L177

Dark pure colors on the TV don't look very clear, they wash out a little, the ESP-Code seems not to take that into account...
Maybe there is a possibility to fix it...?
Or do you have another idea how to fix that in HyperHDR itself (change the saturation of the LEDs, but only for darker pixels)? I did not find a way to change that behavior here...

Sebastian

**Reply by awawa-dev** [reply]
> Where can I find the custom LUT for the Ezcap 320? I have looked everywhere so maybe I am too stupid to find it or I am overseeing it somehow.

It's in the release section: https://github.com/awawa-dev/HyperHDR/releases/tag/v15.0.0.0
It's only for ezcap320 /1080p . ezcap320/4k would required different LUT approach because it's colors are distorted in different way.

> However, with HDR video, too much red is mixed to the blue, even though I set 0,0,255

I had the same problem with blue on ezcap 320 before creating modified LUT table...but Ezcap 320 is specific and unique as it can do partial HDR tone mapping.

The blue led component inside  sk6812 is also very specific..it's very weak. That's why I switched sk6812 neutral white to cold white. Probable it's related to smaller radiators inside the led (compared for example to ws2801) and warmer white component that not exists on other types of led. 

> The only thing I have not found out is that dark colors are too strong. They look very dark in the LED-Visualisation but on the wall they are very pure (too pure). Brighter Colors are OK.

A question: do you use minimal brightness?

> Could you probably insert an option to reduce the color purity for dark colors (mix more white in, keep the brightness)?
If I understand right, this could be done best in the esp8266-code, right? Maybe here (I am not a programmer, so hopefully my logic is right):

The white channel in sk6812 is rather complicated issue and it needs to be treat carefully. Most of algorithms just reduce R,G,B by some amount and write it to W....it has terrible effect for sk6812 neutral and probable warm sk6812 because the 'neutral white' is in fact 'yellow' and W is much stronger than R/B/G.  As you can find in that HyperSerialEsp8266 sketch the proportion for 'neutral white' from experiments are in fact 0xb0/0xb0/0x70 (hexadecimal R/G/B). <br/>I think you selected proper place in the code to do such things anyway ;) For now you can give a try my WLED fork as WLED maybe has similar option.

Added: LUT table for Ezcap 321
![ezcap321_lut](https://user-images.githubusercontent.com/69086569/161051231-24af55a9-4203-4118-8861-772eb4ae744e.png)
